bin/millau/runtime/src/rialto_messages.rs

```diff
@@ -166,12 +168,19 @@ impl TargetHeaderChain<ToRialtoMessagePayload, bp_rialto::AccountId> for Rialto
 	// - id of the lane we prove state of.
 	type MessagesDeliveryProof = ToRialtoMessagesDeliveryProof;
 
-	fn verify_message(payload: &ToRialtoMessagePayload) -> Result<(), Self::Error> {
+	fn verify_message(
```

Comment:
The main idea of splitting `TargetHeaderChain::verify_message` and `LaneMessageVerifier::verify_message` is that you may use single implementation of `TargetHeaderChain` for all lanes (and message-lane pallet instances) that are sending messages to that target chain. So it was supposed to only do things like [that](https://github.com/paritytech/parity-bridges-common/blob/master/primitives/message-lane/src/source_chain.rs#L41). When talking about substrate-based chains, it should check message weight and message size. Everything else is pallet/lane specific and was supposed to be done in `LaneMessageVerifier::verify_message`.

Reply:
Yes, but the dispatcher is not lane specific - it's a property of a target chain. Currently it's even questionable if we need both of them - the split seems unnecessary, cause we do both checks [one after the other](https://github.com/paritytech/parity-bridges-common/blob/3be2b72714641389e2e7638b409e1f26c0f80444/modules/message-lane/src/lib.rs#L254-L266).

Okay, now that I've analyzed this a bit more I think I understand what your point is. We can re-use the same `TargetHeaderChain` for multiple `MessageDelivery` implementations and each message delivery mechanism can be plugged into a different dispatch mechanism.

I'll move the check then.

Reply:
> Yes, but the dispatcher is not lane specific - it's a property of a target chain. Currently it's even questionable if we need both of them - the split seems unnecessary, cause we do both checks one after the other.

Dispatcher may be lane-specific, or even message-specific, it's just our implementation in `bp-runtime-common` assumes that all messages are encoded calls of target chain.